[RFC] Type-directed array literal disambiguation

nojb opened the pull request:
See https://github.com/ocaml/RFCs/pull/46.

The patch modifies the typechecking of array expressions and patterns: if the "expected type" is known to be a `floatarray` (eg because a suitable type annotation has been given by the programmer), then the array literal is taken to denote a constant of that type (how does this interact with principality?). Note that translation into Lambda does not need any adaptations because the code already has logic to translate arrays in a way that handles both normal arrays and flat float arrays:

https://github.com/ocaml/ocaml/blob/48d3c5a065838e41388b18ea8f33d57703ef4337/lambda/translcore.ml#L355

When preparing this PR, I realized that a similar patch had already been discussed in the past, which I had completely forgotten: #9952. However, the approach was slightly different, and the approach here seems to be closer to the one suggested in https://github.com/ocaml/ocaml/pull/9952#issuecomment-742388988.

In any case, I suspect many of the points that were raised in #9952 will again be relevant here. So I consider this PR more of an invitation to (re)open a discussion than a concrete proposal for integration.

An example of the patch in action:

```
OCaml version 5.3.0+dev0-2023-12-22
Enter #help;; for help.

# let x = [|1.3; 4.5|];;
val x : float array = [|1.3; 4.5|]
# x.(1);;
- : float = 4.5
# let x : floatarray = [|1.3; 4.5|];;
val x : floatarray = <abstr>
# Float.Array.get x 1;;
- : float = 4.5
# let f = function [|a; b|] -> a + b;;
val f : int array -> int = <fun>
# f [|1;3|];;
- : int = 4
# let f = function [|a; b|] -> a +. b;;
val f : float array -> float = <fun>
# f [|1.;3.|];;
- : float = 4.
# let f : floatarray -> _ = function [|a; b|] -> a +. b;;
val f : floatarray -> float = <fun>
# f [|1.;3.|];;
- : float = 4.
```

nojb committed:
Type-directed array literal disambiguation

goldfirere commented:
The code is delightfully simple. Thank you! I think there are a few ways in which it could be improved -- notably, by checking whether an expected type is principally known. Also, I worry about the use of `Typeopt` here. That looks through `private` and `[@@unboxed]` types, which might mean we e.g. construct a value of a synonym of a `floatarray` that maintains an invariant without checking the invariant.

I'm confident the above issues can be resolved. So the bigger question is: once the implementation has settled, should we merge? I'd like to, but we should form some consensus on this point.

In the meantime, do you think you could add some tests to show this in action? That might help to crystallize our understanding. Or, if your interest in this was just as a small mockup, let me know and we can look into taking this over.

nojb commented:
> notably, by checking whether an expected type is principally known.

I gave it a try, but could not figure out how to do it (I am not sure what this means exactly, in fact). Do you have any pointers?

> In the meantime, do you think you could add some tests to show this in action?

Yes, I will add some tests soon. Thanks!

goldfirere commented:
A type is known principally if the knowledge of that type is completely predictable. Maybe it would be better to say "known predictably" or "specified by the user". A type annotation can cause a type to be known principally, or being an argument to a function whose argument type is known principally.

The good news is that it's very easy to check this property: just use `is_principal` (from `Ctype`). And it seems we still allow disambiguation even with not-principally-known types, but we warn when doing so. Use `warn_non_principal`.

And instead of `Typeopt`, I recommend `expand_head`, which should have the behavior you're looking for.

Thanks!

nojb committed:
is_principal, expand_head

nojb committed:
Add test

nojb commented:
Thanks for the pointers @goldfirere!

> just use `is_principal` (from `Ctype`).

For future reference, this is apparently referencing the following function of the JS fork:

https://github.com/ocaml-flambda/flambda-backend/blob/bca81e9f87b7833089d16bbfd694fc8be36dacf5/ocaml/typing/ctype.mli#L631-L633

which I cherry-picked to this PR.

> And instead of `Typeopt`, I recommend `expand_head`, which should have the behavior you're looking for.

Thanks, I did that. I am not sure if the way I check whether the expected type "is" `floatarray` is the most adequate one, though.

> > In the meantime, do you think you could add some tests to show this in action?
> 
> Yes, I will add some tests soon. Thanks!

I added a few tests to get started (including one that exercises the principality aspect).

gasche commented:
`is_principal` is available in trunk, but it is in Typecore. You can move it to Ctype as in the JS fork, but then you should remove the one in Typecore.

nojb committed:
Remove duplicated function is_principal

nojb commented:
> `is_principal` is available in trunk, but it is in Typecore. You can move it to Ctype as in the JS fork, but then you should remove the one in Typecore.

Thanks, I had missed it; I removed the one I had added to `Ctype`.

goldfirere commented on typing/typecore.ml:
```diff
@@ -975,11 +975,18 @@ let solve_Ppat_record_field ~refine loc penv label label_lid record_ty =
+     | Tconstr (path, [], _) -> Path.same path Predef.path_floatarray
+     | _ -> false
+  then
+    Predef.type_float, Predef.type_floatarray
```
I don't think this needs to return the `type_floatarray` here. The existing treatment of the pattern type in `type_pat_aux` should work.

nojb replied:
Thanks, fixed.

goldfirere commented on typing/typecore.ml:
```diff
@@ -975,11 +975,18 @@ let solve_Ppat_record_field ~refine loc penv label label_lid record_ty =
+     | Tconstr (path, [], _) -> Path.same path Predef.path_floatarray
+     | _ -> false
+  then
+    Predef.type_float, Predef.type_floatarray
```
Someone more expert in these matters than me may say otherwise, but I think you want `generic_instance Predef.type_float` for the element type. This seems to match the fact that the general case calls `newgenvar`.

Generally speaking, you don't want to use types from `Predef` directly; you usually want to take an `instance` or `generic_instance` of them. Using the type directly (without `instance`) means taking the master copy and putting it in a structure. If we then learn more about the type (say by a GADT pattern-match), we might influence that master copy.

nojb replied:
I ended up using `instance Predef.type_float` (I guess that for parameter-less types there is no difference). This matches what is done elsewhere in `Typecore`, eg `type_constant`.

goldfirere replied:
It surprised me, too, but there is a difference, actually. Both `instance` and `generic_instance` copy the type. `instance` additionally changes all levels that are `generic_level` to be the current level (that is, the result of `get_current_level ()`). We consider a type whose level is `generic_level` to be principally known: either this type came from a user-written type annotation, or inference on the type has been completed (for the type of a let-bound `x`) and the type has been generalized. So `instance Predef.type_float` produces a `float` that is not principally known (and couldn't be used in type-directed disambiguation, say) while `generic_instance Predef.type_float` *can* be used for type-directed disambiguation.

I see `Typecore.type_constant` uses `instance`. I'm not actually sure why that is: If I see that something is `1`, its type is definitely `int`; saying `(1 : int)` shouldn't be different from `1`. So I'm not exactly sure what's best here; we should wait for @garrigue.

nojb replied:
OK, I left `instance` for now until the issue is clarified. Thanks for the explanation by the way, it is appreciated!

garrigue replied:
@goldfirere Type synthesized by `type_exp` are supposed to have a level no higher than `current_level`, which reflects the fact than syntax-directed rules for Hindley-Milner have only monotypes on the right-hand side. To make easier to follow this invariant, most functions return type nodes at `current_level`. Of course, one could start from `generic_level` and then use `update_level` to get back to `current_level`, but this introduces extra costs.

goldfirere replied:
I see what you're saying here, but the existence of `generic_level` is additionally used to affect principality of type inference. If I have an expression `5`, I would expect its type `int` to be principally known. This would suggest using `generic_instance` instead of `instance`. But now that I think about it, I can't find a way in which this would change any user-facing behavior, so maybe this isn't a point that needs debating...

goldfirere commented on typing/typecore.ml:
```diff
@@ -3865,16 +3872,27 @@ and type_expect_
-        unify_exp_types loc env to_unify (generic_instance ty_expected));
+      let ty_elt, exp_type =
+        let ty_expected = generic_instance ty_expected in
+        if is_principal ty_expected &&
+           match get_desc (expand_head env ty_expected) with
+           | Tconstr (path, [], _) -> Path.same path Predef.path_floatarray
+           | _ -> false
```
This little bit is repeated. Maybe abstract into a function?

nojb replied:
Good idea, done. Thanks.

goldfirere commented on typing/typecore.ml:
```diff
@@ -3865,16 +3872,27 @@ and type_expect_
+           | Tconstr (path, [], _) -> Path.same path Predef.path_floatarray
+           | _ -> false
+        then
+          Predef.type_float, Predef.type_floatarray
```
Same here as above: no need to return this here. The `ty_expected` should work fine.

nojb replied:
Fixed as well, thanks.

goldfirere reviewed (commented):
Implementation looks pretty good to me, modulo some suggestions I've made here.

But do we want this patch? I'm not the one to decide such things.

Also, this would need a user's manual update before merging.

nojb commented:
> modulo some suggestions I've made here.

Thanks for the review @goldfirere! I will tweak the patch as suggested shortly.

> But do we want this patch? I'm not the one to decide such things.

Some names off the top of my head whose opinion would be welcome: @yallop @garrigue @lpw25 @gasche

> this would need a user's manual update before merging.

Indeed. However, I will wait to see if the discussion advances a bit more before doing this part.

garrigue commented:
I do want to look into this, in particular the principality part, but I see that I can be slow. So in the end you don't need to wait for me too long.

nojb committed:
Simplify

nojb committed:
Share code

nojb committed:
Call instance on Predef.type_float

goldfirere commented on typing/typecore.ml:
```diff
@@ -975,11 +980,14 @@ let solve_Ppat_record_field ~refine loc penv label label_lid record_ty =
-  unify_pat_types_refine ~refine
-    loc env (Predef.type_array ty_elt) expected_ty;
-  ty_elt
+  if is_principal expected_ty && is_floatarray_type !!env expected_ty then
+    instance Predef.type_float
```
Sorry to nitpick, but this bit is *still* repeated. :)

Maybe something like

```
(* Returns [Some ty_elt] if the expected type can be used for type-directed disambiguation
   of the array literal. *)
let disambiguate_array_literal env expected_ty =
  if is_principal expected_ty then begin
    if is_floatarray_type env expected_ty then Some (instance Predef.type_float) else
    None
  end else None
```

and then this can be used here and for expressions. The structure is designed to be easily extensible with new types that might participate in this.

nojb replied:
Good point, I amended as suggested. Thanks!

goldfirere commented:
I don't think this is particularly pressing, so I'm happy to wait for @garrigue. I will bring this up at our next meeting, if you don't get a chance to get to it beforehand.

nojb committed:
Share code

goldfirere commented:
I'm happy with the implementation, pending more expert review, which we understand may take some time to get.

goldfirere commented:
And I forgot to say: Thanks, @nojb for doing this!

yallop commented:
This looks good to me, but I think we should also add top-level printing support, so that we see

```
# let x : floatarray = [| 3.; 4. |];;
val x : floatarray = [|3.; 4.|]
```

rather than

```
# let x : floatarray = [| 3.; 4. |];;
val x : floatarray = <abstr>
```

I've proposed an implementation here: https://github.com/nojb/ocaml/pull/13

yallop committed:
Add support for top-level floatarray printing. (#13)

nojb commented:
> I've proposed an implementation here: [nojb#13](https://github.com/nojb/ocaml/pull/13)

This is now merged into the PR. Thanks again!

alainfrisch commented on typing/typecore.ml:
```diff
@@ -381,6 +386,17 @@ let extract_label_names env ty =
+   disambiguation of an array literal. *)
+let disambiguate_array_literal env expected_ty =
+  if is_principal expected_ty then
+    if is_floatarray_type env expected_ty then
```
I assume this is written like that (and not `if is_principal expected_ty && is_floatarray_type env expected_ty then Some ... else None`) to support, later, other kinds of arrays. Is that right?

In particular, it could already be useful to support `bytes` (and possibly `string`) literals. Would that require a lot of extra changes in the backend?

alainfrisch commented:
I like the feature!

I believe that disambiguating get/set operations would be even more useful (and coherent with records). Do we foresee extra difficulties, compared to this PR, both in terms of implementation and acceptability of the feature?

garrigue commented:
I see no problem on the principality side.
To follow-up on the discussion in #9952, which I did not attend, (local) ambiguity and principality are two different problems. Local ambiguity says that there is a subexpression that could be given two different meanings, ignoring the context. Principality says that there exists a type system such that all programs have at most one principal type, or no type at all.
What you do here is fine for principality, since either there is positive information that this is a `floatarray`, that should be available in any derivation, or there is none, and you just revert to the default case, which is incompatible with `floatarray`. If there is partial information, you fail, which is again ok, since the same type system could be made to not accept such cases.

This said, in ocaml we usual try to make the principal mode compatible with the default mode, i.e. if the default mode succeeds, the principal mode should ideally succeed with a principality warning. So I believe the order of the checks should be inverted: you should first check whether it is a `floatarray`, and if it is, then check for principality and emit a warning if needed. This would be more helpful to programmers.

Note that the above comment on principality assumes that expected types only come from type annotations and constructors. In practice ocaml also propagates expected types from function to their arguments, which is not strictly principal, but only principal modulo the maximality of derivations, i.e. we need to disallow taking derivations which instantiate types more than needed, as they could end up being more "informative". Some people call this "restriction to principal derivations", which I find rather strange, as it makes the definition cyclic.

nojb committed:
Warn if type disambiguation is not principal

nojb commented:
> So I believe the order of the checks should be inverted:

Thanks for the review @garrigue! I committed your suggestion in https://github.com/ocaml/ocaml/pull/13340/commits/ce706fbab2e7484892424554925b37dce0c2ff93 (the effect can be seen in the one of the tests).

gasche commented:
We now have a consensus among language maintainers to merge this feature, and the implementation has been reviewed for correctness. Great! But before I click "merge", I would like to ask about documentation: how are users supposed to find out about this feature? I think it would be nice if you could add some documentation for the PR, either in the user manual (if you find a good place to put it) or, at least, in the header documentation of the Float.Array module.

(It would be nice if the documentation included examples where type inference works as intended, and also examples where it does not and adding an annotation is useful.)

gasche commented:
(Also check-typo: no Changes, and line too long at `./typing/typecore.ml:3903.81`)

nojb commented:
Thanks @gasche for working to achieve consensus around this feature!

> I think it would be nice if you could add some documentation for the PR, either in the user manual (if you find a good place to put it) or, at least, in the header documentation of the Float.Array module.

I will see to add something to the manual.

> (Also check-typo:

Will fix, thanks.

nojb committed:
Merge remote-tracking branch 'upstream/trunk' into type_directed_array_literals

nojb committed:
Add doc to manual 'Language extensions'

nojb committed:
Changes

nojb committed:
check-typo

nojb commented:
I pushed some documentation in a new section of the "Language extensions" chapter (comments welcome), updated `Changes` and fixed the `check-typo` warnings.

goldfirere commented on manual/src/refman/extensions/arrayliterals.etex:
This sentence is redundant to me — I would just drop it.

nojb replied:
Done, thanks.

goldfirere commented on manual/src/refman/extensions/arrayliterals.etex:
```diff
@@ -0,0 +1,74 @@
+
+\begin{caml_example}{verbatim}
+let x = [|42.|];;
+\end{caml_example}
```
I would explicitly say that `x` gets the type `float array` here.

Octachron replied:
I am not sure if this is needed since the `caml_example` environment will show the toplevel output:
```ocaml
val x : float array = [|42.|]
```

nojb replied:
Even though it is contained in the evaluation output, I still added the sentence to the documentation; the repetition did not feel superfluous.

goldfirere commented on manual/src/refman/extensions/arrayliterals.etex:
Drop “Indeed” here

nojb replied:
Done, thanks.

goldfirere commented on manual/src/refman/extensions/arrayliterals.etex:
I would say somewhere that the rules here are exactly the same as for constructor disambiguation.

Octachron replied:
I agree that the description should mention type-directed disambiguation for constructors and records.

nojb replied:
I added a mention of this, with a reference to the section in question.

goldfirere reviewed (commented):
I’m not experienced with the usual formatting, etc., of the manual, but I’ve read the text and offered a few small comments. As always, thanks for this!

Octachron commented on manual/src/refman/extensions/arrayliterals.etex:
```diff
@@ -0,0 +1,74 @@
+(Introduced in OCaml 5.4)
+
+Since OCaml 5.4, array literal syntax @"[|" e1 ';' e2 ';' ... ';' eN "|]"@ can
+be used to denote values of type "'a array" as well as "floatarray", both in
```
I would reverse the enumeration: 

> Since OCaml 5.4 ... can be used to denote values of type  "floatarray" as well as "'a array", both in

to put the 5.4 novelty first.

nojb replied:
Done, thanks.

Octachron commented on manual/src/refman/extensions/arrayliterals.etex:
```diff
@@ -0,0 +1,74 @@
+
+The compiler will use contextual information to infer the type of the array
+literal. However, in the absence of any specific contextual information, it will
+be assumed that the array literal is of type "'a array":
```
Since the "contextual information" that the typechecker is using is the expected type, I find this description a bit too circular.

Maybe 

> The compiler matches  the expected type of the expression or pattern with the type of the literal. In the absence of an expected type, array literals are assumed to be of type "'a array":

nojb replied:
Amended as suggested, thanks.

Octachron commented on manual/src/refman/extensions/arrayliterals.etex:
I have the impression that this sentence is redundant once we state clearly that the compiler is using the expected type. I would keep the examples but move them before the "defaulting to 'a array".

nojb replied:
I removed the sentence (as suggested by @goldfirere as well).

Octachron commented on manual/src/refman/extensions/arrayliterals.etex:
I am not sure we need this section, and the term `any type information that it learns from the context` is too imprecise and is contradicted by your next example.

nojb replied:
I removed this paragraph. Thanks.

Octachron commented on manual/src/refman/extensions/arrayliterals.etex:
This would be the right place to mention the `non-principal-disambiguation` warning.

nojb replied:
I added a mention of the warning just after the example.

Octachron commented on manual/src/refman/extensions/arrayliterals.etex:
```diff
@@ -0,0 +1,74 @@
+(Introduced in OCaml 5.4)
```
You are missing the forward links in `expr.etex` and `patterns.etex`.

nojb replied:
Fixed, thanks.

nojb committed:
Review

nojb commented:
Thanks @goldfirere and @Octachron for your suggestions concerning the documentation. I pushed a new commit that should address all points; I also reworded a bit the last paragraph which illustrates the non-principal case. I would appreciate any further comments you may have. Thanks!

goldfirere commented:
Looks good to me!

gasche reviewed (approved):
Approving based on the existing reviews, consensus on the feature, and documentation. Thanks!

I am planning to let it float for a day or two, forget about it, then merge.
